Add the feature to make the promotion rule be used once

hparfr opened the pull request:
@PierrickBrun

hparfr committed:
Add the feature to make the promotion rule be used once

Commit:
[REF] rename currency_id for more integration with ORM

Commit:
[IMP] add valid_once and max_budget restriction_types

Now, usage restrictions are computed for confirmed sale_orders only:
You can are 2 draft SO with promotions -> when the first one is
validated, the promotions are removed from the second

hparfr commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -103,6 +107,49 @@ class SalePromotionRule(models.Model):
+            )
+            if lines:
+                if rec.discount_type == "amount_tax_included":
+                    rec.budget_spent = -sum(lines.mapped("price_total"))
```
it should not be negative

PierrickBrun replied:
Since it's a discount, price_total is always negative, so I negate it again to have a positive value. Maybe I should use `abs()` ?

hparfr replied:
It's named budget_spent. So it should be positive.

PierrickBrun replied:
Yes, `budget_spent` is positive, this is `lines.price_total` that is negative

hparfr replied:
Ok I get it.

hparfr commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -103,6 +107,49 @@ class SalePromotionRule(models.Model):
+                    rec.budget_spent = -sum(lines.mapped("price_total"))
+                else:
+                    rec.budget_spent = -sum(
+                        lines.mapped("untaxed_amount_to_invoice")
```
It should be something else because untaxed_amount_to_invoice depends if the invoice has been generated or not (and if the goods have been delievered?).

hparfr replied:
it's price_subtotal

hparfr commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -51,6 +51,13 @@ def _compute_applied_promotion_rule_ids(self):
+    def action_confirm(self):
+        res = super().action_confirm()
+        for order in self:
+            order.applied_promotion_rule_ids.check_used()
```
why not self.applied_promotion_rule_ids.check_used() ?

Commit:
!fixup misc.

hparfr committed:
Add some fields in the view

Commit:
[FIX] maximum_budget will not reduce price under minimal_amount

Commit:
!fixup unlink line if there is no sale order

Commit:
!fixup fix amount recalculation take excluded lines into account

Commit:
[IMP] Allow to override sale_order_line used condition

Commit:
[FIX] set original discount (from pricelist)
on so line when removing promotions

Commit:
[IMP] do not remove promotion_lines but modify them instead

Commit:
[FIX] rule.used computation

Commit:
[FIX] Can't reduce order price under rule minimal amount

Commit:
[CHG] price_unit is positive and qty is -1
This is needed for facturx compatibility

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (3 of 3 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_archive
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_archive/pt/

eLBati committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (8 of 8 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_line_sequence/it/

sergiocorato committed:
[12.0][FIX] update hooked function to original function create_invoices
see https://github.com/odoo/odoo/commit/ff4f1eafd222555b6c13503a9d7253403899bcee

OCA-git-bot committed:
Merge PR #1605 into 12.0

Signed-off-by rousseldenis

oca-travis committed:
[UPD] Update sale_order_action_invoice_create_hook.pot

OCA-git-bot committed:
sale_order_action_invoice_create_hook 12.0.1.0.1

OCA-git-bot committed:
[UPD] addons table in README.md

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_action_invoice_create_hook
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_action_invoice_create_hook/

oca-travis committed:
[UPD] Update sale_order_action_invoice_create_hook.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_action_invoice_create_hook
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_action_invoice_create_hook/

Commit:
Translated using Weblate (Italian)

Currently translated at 100.0% (32 of 32 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_type
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_type/it/

ypapouin committed:
Added translation using Weblate (French)

ypapouin committed:
Translated using Weblate (French)

Currently translated at 100.0% (20 of 20 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_tag
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_tag/fr/

OCA-git-bot committed:
Merge PR #1117 into 12.0

Signed-off-by rousseldenis

oca-travis committed:
[UPD] Update sale_promotion_rule.pot

OCA-git-bot committed:
[UPD] addons table in README.md

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
sale_promotion_rule 12.0.1.0.1

OCA-git-bot committed:
[ADD] setup.py

OCA-git-bot committed:
[UPD] addons table in README.md

bealdav committed:
Added translation using Weblate (French)
